diff -Nru linux-5.5.4.new/drivers/cartesi/Kconfig linux-5.5.4.newer/drivers/cartesi/Kconfig
--- linux-5.5.4.new/drivers/cartesi/Kconfig	2020-06-18 15:44:03.000000000 -0300
+++ linux-5.5.4.newer/drivers/cartesi/Kconfig	2020-06-18 16:02:23.000000000 -0300
@@ -14,4 +14,11 @@
     help
         Cartesi Machine Yield Device
 
+config CARTESI_HALT_ON_PANIC
+    bool "Halt machine on kernel panic"
+    depends on CARTESI_MACHINE
+    default y
+    help
+        Halt machine on kernel panic
+
 endmenu
diff -Nru linux-5.5.4.new/kernel/panic.c linux-5.5.4.newer/kernel/panic.c
--- linux-5.5.4.new/kernel/panic.c	2020-06-18 14:32:12.000000000 -0300
+++ linux-5.5.4.newer/kernel/panic.c	2020-06-18 16:02:24.000000000 -0300
@@ -333,6 +333,11 @@
 			 "twice on console to return to the boot prom\n");
 	}
 #endif
+
+#ifdef CONFIG_CARTESI_HALT_ON_PANIC
+machine_halt();
+#endif
+
 #if defined(CONFIG_S390)
 	disabled_wait();
 #endif
